### PR TITLE
Error recovery configuration

### PR DESCRIPTION
**Pull request by joesus**
## Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I've completed the [Contributor License Agreement](https://developers.facebook.com/opensource/cla)
- [x] I've ensured that all existing tests pass and added tests (when/where necessary)
- [x] I've ensured that my code lints properly: (`swiftlint && swiftlint autocorrect --format`)
- [x] I've updated the documentation (when/where necessary) and [Changelog](CHANGELOG.md) (when/where necessary)
- [x] I've added the proper label to this pull request (e.g. `bug` for bug fixes)

## Pull Request Details

Introduces error recovery configuration struct and concept of remote error configuration.
This introduces a pattern to build Canonical data models from Remote (server-side) data.
This pull is for the first part of the equation, making sure we can create a representation of the remote data that we can use for building an ErrorConfiguration. This prevents our canonical model from being too dependent on the actual server response.

## Test Plan

Should build without issue. Tests should pass.

### CODE REVIEW

File: Sources/Core/RemoteErrorRecoveryCodes.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/// Used for creating a `RemoteErrorRecoveryConfiguration`
+struct RemoteErrorRecoveryCodes: Codable, Equatable {
+  let primaryCode: Int
+  let subcodes: [Int]
+
+  init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: Keys.self)
+    primaryCode = try container.decode(Int.self, forKey: .code)
+
+    if var subcodesContainer = try? container.nestedUnkeyedContainer(forKey: .subcodes) {
```
Comment (codytwinton) [inline Sources/Core/RemoteErrorRecoveryCodes.swift]: switch?

File: Sources/Core/RemoteErrorRecoveryConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    let name = try container.decode(String.self, forKey: .name)
+    guard !name.isEmpty else {
+      throw RemoteErrorConfigurationDecodingError.emptyName
+    }
+
+    self.name = name
+
+    while !itemsContainer.isAtEnd {
+      if let item = try? itemsContainer.decode(RemoteErrorRecoveryCodes.self) {
```
Comment (codytwinton) [inline Sources/Core/RemoteErrorRecoveryConfiguration.swift]: Switch?

File: Sources/Core/RemoteErrorRecoveryCodes.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      var subcodes: [Int] = []
+      while !subcodesContainer.isAtEnd {
+        if let code = try? subcodesContainer.decode(Int.self) {
```
Comment (codytwinton) [inline Sources/Core/RemoteErrorRecoveryCodes.swift]: switch

File: Sources/Core/RemoteErrorRecoveryConfigurationList.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+/// Used for creating an `ErrorConfiguration`
+struct RemoteErrorRecoveryConfigurationList: Codable {
+  let configurations: [RemoteErrorRecoveryConfiguration]
+
+  init(from decoder: Decoder) throws {
+    var container = try decoder.unkeyedContainer()
+    var configurations: [RemoteErrorRecoveryConfiguration] = []
+
+    while !container.isAtEnd {
+      if let item = try? container.decode(RemoteErrorRecoveryConfiguration.self) {
```
Comment (codytwinton) [inline Sources/Core/RemoteErrorRecoveryConfigurationList.swift]: switch

File: Sources/FacebookCoreTests/Sample Data/SampleRawRemoteConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    static func missing(_ key: String) -> Data {
+      var temp = validDictionary
+      temp.removeValue(forKey: key)
+      return try! JSONSerialization.data(withJSONObject: temp, options: [])
+    }
+  }
+}
+
+// TODO:
```
Comment (codytwinton) [inline Sources/FacebookCoreTests/Sample Data/SampleRawRemoteConfiguration.swift]: Remove?

File: Sources/Core/RemoteErrorRecoveryCodes.swift (hunk 1 above)
Comment (joesus) [reply Sources/Core/RemoteErrorRecoveryCodes.swift]: done!

File: Sources/Core/RemoteErrorRecoveryCodes.swift (hunk 3 above)
Comment (joesus) [reply Sources/Core/RemoteErrorRecoveryCodes.swift]: done!

File: Sources/Core/RemoteErrorRecoveryConfigurationList.swift (hunk 4 above)
Comment (joesus) [reply Sources/Core/RemoteErrorRecoveryConfigurationList.swift]: done

**Pull request merged**